### PR TITLE
remove unsafety from util/hash.rs and util/bigint/uint.rs

### PR DESCRIPTION
**Pull request by rphmeier**
the copy stuff was fine, but some of the uses of `mem::uninitialized` were dubious.

### CODE REVIEW

File: util/src/hash.rs [hunk 1]
```diff
@@ -548,51 +527,44 @@ impl<'a> From<&'a H256> for U256 {
...
 }
+
 /*
 impl<'a> From<&'a H256> for Address {
 	fn from(value: &'a H256) -> Address {
-		unsafe {
-			let mut ret: Address = ::std::mem::uninitialized();
-			::std::ptr::copy(value.as_ptr().offset(12), ret.as_mut_ptr(), 20);
+			let mut ret = Address::new();
+			ret.0.copy_from_slice(&value[12..32]);
```
Comment (rphmeier) [inline util/src/hash.rs]: side note: anyone know why this is commented out, and if it could be uncommented?

File: util/src/hash.rs [hunk 2]
```diff
@@ -150,10 +150,8 @@ macro_rules! impl_hash {
 			}
 
 			fn copy_to(&self, dest: &mut[u8]) {
-				unsafe {
 					let min = ::std::cmp::min($size, dest.len());
-					::std::ptr::copy(self.0.as_ptr(), dest.as_mut_ptr(), min);
-				}
+					dest[..min].copy_from_slice(&self.0[..min]);
```
Comment (tomusdrw) [inline util/src/hash.rs]: `<<`

File: util/src/hash.rs [hunk 3]
```diff
@@ -163,10 +161,7 @@ macro_rules! impl_hash {
 				// impl |= instead
 				// TODO: that's done now!
```
Comment (tomusdrw) [inline util/src/hash.rs]: Is the todo and comment still valid?

File: util/src/hash.rs (hunk 1 above)
Comment (tomusdrw) [inline util/src/hash.rs]: IMHO it should be removed. H256 is `Copy` anyway (I guess same for `From<&Address>` ?)

File: util/src/hash.rs [hunk 4]
```diff
@@ -548,51 +527,44 @@ impl<'a> From<&'a H256> for U256 {
...
 
 impl<'a> From<&'a Address> for H256 {
 	fn from(value: &'a Address) -> H256 {
-		unsafe {
-			let mut ret = H256::new();
-			::std::ptr::copy(value.as_ptr(), ret.as_mut_ptr().offset(12), 20);
-			ret
-		}
+		let mut ret = H256::new();
+		ret.0[12..32].copy_from_slice(&value);
```
Comment (tomusdrw) [inline util/src/hash.rs]: `&` not needed here?

File: util/src/hash.rs (hunk 3 above)
Comment (rphmeier) [inline util/src/hash.rs]: not sure if it ever was, at least how i'm reading it.
we don't have any `impl BitOrAssign` for hash types, and it also isn't implemented for slices.

File: util/src/hash.rs (hunk 4 above)
Comment (rphmeier) [inline util/src/hash.rs]: It is necessary for the deref coercion. However, it was unneeded in the `impl<'a> From<&'a Address> for H256` version -- I have addressed this.

**Comment by rphmeier** [comment]
gonna try and take out as much unsafety from U256 as possible as well -- won't be able to get rid of all of it, since there's a fair bit of inline asm, but probably a decent chunk.

File: util/src/hash.rs (hunk 1 above)
Comment (rphmeier) [inline util/src/hash.rs]: Hm, it can't hurt much to have both impls -- I would opt to keep it as long as everything compiles with it.

**Pull request merged**